bee-tangle/src/urts.rs

```diff
@@ -0,0 +1,197 @@
+// gets semi-lazy.
+const OMRSI_DELTA: u32 = 13;
+// M: the maximum allowed delta value between OMRSI of a given message in relation to the current SMI before it
+// gets lazy.
+/// The maximum permitted difference between a message's OMRSI and the current SMI before it is considered lazy.
+pub const BELOW_MAX_DEPTH: u32 = 15;
+// If the amount of non-lazy tips exceed this limit, remove the parent(s) of the inserted tip to compensate for the
+// excess. This rule helps to reduce the amount of tips in the network.
+const MAX_LIMIT_NON_LAZY: u8 = 100;
+// The maximum time a tip remains in the tip pool after having the first child.
+// This rule helps to widen the tangle.
+const MAX_AGE_SECONDS_AFTER_FIRST_CHILD: u8 = 3;
+// The maximum amount of children a tip is allowed to have before the tip is removed from the tip pool. This rule is
+// used to widen the cone of the tangle.
+const MAX_NUM_CHILDREN: u8 = 2;
+
+#[derive(Default)]
+struct TipMetadata {
+    children: HashSet<MessageId>,
+    time_first_child: Option<Instant>,
+}
+
+impl TipMetadata {
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct UrtsTipPool {
+    tips: HashMap<MessageId, TipMetadata>,
+    non_lazy_tips: HashSet<MessageId>,
+}
+
+impl UrtsTipPool {
+    pub(crate) fn non_lazy_tips(&self) -> &HashSet<MessageId> {
+        &self.non_lazy_tips
+    }
+
+    pub(crate) async fn insert<B: StorageBackend>(
+        &mut self,
+        tangle: &MsTangle<B>,
+        message_id: MessageId,
+        parents: Vec<MessageId>,
+    ) {
+        if let Score::NonLazy = self.tip_score::<B>(tangle, &message_id).await {
+            self.non_lazy_tips.insert(message_id);
+            self.tips.insert(message_id, TipMetadata::new());
+            for parent in parents.iter() {
+                self.add_child(*parent, message_id);
+                self.check_retention_rules_for_parent(parent);
+            }
+        }
+    }
+
+    fn add_child(&mut self, parent: MessageId, child: MessageId) {
+        match self.tips.entry(parent) {
+            Entry::Occupied(mut entry) => {
+                let metadata = entry.get_mut();
+                metadata.children.insert(child);
+                if metadata.time_first_child == None {
+                    metadata.time_first_child = Some(Instant::now());
+                }
+            }
+            Entry::Vacant(entry) => {
+                let mut metadata = TipMetadata::new();
+                metadata.children.insert(child);
+                metadata.time_first_child = Some(Instant::now());
+                entry.insert(metadata);
+            }
+        }
+    }
+
+    fn check_retention_rules_for_parent(&mut self, parent: &MessageId) {
+        if self.non_lazy_tips.len() > MAX_LIMIT_NON_LAZY as usize
+            || self.tips.get(parent).unwrap().children.len() as u8 > MAX_NUM_CHILDREN
```

Comment:
could we `self.tip.get(parent)` once? also, why is it ok to unwrap this?

Reply:
Line 73 makes sure the parent is available, so it's safe to unwrap()
We need a comment here

Reply:
Cheers